src/End.sol

```diff
@@ -0,0 +1,335 @@
+        z = x * WAD / y;
+    }
+
+    // --- Administration ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        require(live == 1, "End/not-live");
+        if (what == "vow") vow = data;
+        else if (what == "pot") pot = PotLike(data);
+        else if (what == "spot") spot = SpotLike(data);
+        else if (what == "cure") cure = CureLike(data);
+        else if (what == "claim") claim = ClaimLike(data);
+        else revert("End/file-unrecognized-param");
+        emit File(what, data);
+    }
+    function file(bytes32 what, uint256 data) external auth {
+        require(live == 1, "End/not-live");
+        if (what == "wait") wait = data;
+        else revert("End/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    // --- Settlement ---
+    function cage() external auth {
+        require(live == 1, "End/not-live");
+        live = 0;
+        when = block.timestamp;
+        vat.cage();
+        spot.cage();
+        pot.cage();
+        cure.cage();
+        emit Cage();
+    }
+
+    function cage(bytes32 ilk) external {
+        require(live == 0, "End/still-live");
+        require(tag[ilk] == 0, "End/tag-ilk-already-defined");
+        (Art[ilk],,,,) = vat.ilks(ilk);
+        (PipLike pip,) = spot.ilks(ilk);
+        // par is a ray, pip returns a wad
+        tag[ilk] = wdiv(spot.par(), uint256(pip.read()));
+        emit Cage(ilk);
+    }
+
+    function skim(bytes32 ilk, address urn) external {
+        require(tag[ilk] != 0, "End/tag-ilk-not-defined");
+        (, uint256 rate,,,) = vat.ilks(ilk);
+        (uint256 ink, uint256 art) = vat.urns(ilk, urn);
+
+        uint256 owe = rmul(rmul(art, rate), tag[ilk]);
+        uint256 wad = min(ink, owe);
+        gap[ilk] = gap[ilk] + (owe - wad);
+
+        require(wad <= 2**255 && art <= 2**255, "End/overflow");
+        vat.grab(ilk, urn, address(this), vow, -int256(wad), -int256(art));
+        emit Skim(ilk, urn, wad, art);
+    }
+
+    function free(bytes32 ilk) external {
+        require(live == 0, "End/still-live");
+        (uint256 ink, uint256 art) = vat.urns(ilk, msg.sender);
+        require(art == 0, "End/art-not-zero");
+        require(ink <= 2**255, "End/overflow");
+        vat.grab(ilk, msg.sender, msg.sender, vow, -int256(ink), 0);
+        emit Free(ilk, msg.sender, ink);
+    }
+
+    function thaw() external {
+        require(live == 0, "End/still-live");
+        require(debt == 0, "End/debt-not-zero");
+        require(vat.dai(vow) == 0, "End/surplus-not-zero");
+        require(block.timestamp >= when + wait, "End/wait-not-finished");
+        debt = vat.debt() - cure.tell();
+        claim.mint(address(this), debt);
+        claim.approve(vow, type(uint256).max);
```

Comment:
Why do we need to approve the `vow` here for the claim token? I don't see Vow code in this PR or repo.

Reply:
`vow == DomainGuest` in `xdomain-dss`. It needs approval because it transfers out the claim tokens when users `DomainGuest.exit()`.

Reply:
We probably want to put some comment about what the `vow` represents for `xdomain-dss`.

Reply:
> It needs approval because it transfers out the claim tokens when users DomainGuest.exit().

I'm not seeing this?
https://github.com/makerdao/dss-bridge/blob/v1/src/DomainGuest.sol#L317

Reply:
It exists at https://github.com/makerdao/dss-bridge/blob/v1/src/DomainGuest.sol#L327

<img width="652" alt="Screen Shot 2022-11-08 at 7 48 18 AM" src="https://user-images.githubusercontent.com/588921/200567776-e9ba2aed-5449-4e4d-847f-d5558d7876c0.png">

Reply:
Hmmm weird think my GitHub browser caching was messed up or something...my link now points to an empty line and definitely didn't when I commented.